### PR TITLE
Add middleware stack

### PR DESCRIPTION
**Pull request by SimonNumberTwo**
# Description
This PR adds the middleware stack functionality, now the user is able to add middleware components which mutate the request before sending it. So for example adding header fields or authentication method or anything necessary. 

This system is highly flexible and extendable so it fits all needs and does not need a framework adjustment.

This PR also adds a default implementation for the `HeaderMiddlewareComponent` as well as the `AuthenticationMiddlewareComponent` which can be used by the programmer using this framework. More default implementations will come. But also keep in mind that the programmer him-/herself can use their own implementations.

See also #12 for more details

### CODE REVIEW

File: Sources/Jetworking/Client.swift [hunk 1]
```diff
@@ -69,55 +69,16 @@ public final class Client {
...
-            return [authorizationHeaderKey: authString]
-
-        case let .bearerToken(token):
-            return [authorizationHeaderKey: "Bearer \(token)"]
-
-        case let .custom(headerKey, headerValue):
-            return [headerKey: headerValue]
+        var request = URLRequest(url: clientConfiguration.baseURL.appendingPathComponent(pathComponent), httpMethod: httpMethod)
+        
+        clientConfiguration.middlewareComponents.forEach { middlewareComponent in
```
[the remark this answers is not in the document]
Comment (SimonNumberTwo) [reply Sources/Jetworking/Client.swift:74]: Done.

File: Sources/Jetworking/Middleware/Defaults/AuthenticationMiddlewareComponent.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import Foundation
+
+final class AuthenticationMiddlewareComponent: MiddlewareComponent {
```
[the remark this answers is not in the document]
Comment (SimonNumberTwo) [reply Sources/Jetworking/Middleware/Defaults/AuthenticationMiddlewareComponent.swift:3]: The idea behind making it final is to prevent overwriting them. The programmer using this framework shall create their own implementation and conform to `MiddlewareComponent` instead of `AuthenticationMiddlewareComponent`. What I could think of is to move the `AuthenticationMethod` enum into this component to be more precise.

What do you think of this?

File: Sources/Jetworking/Middleware/Defaults/HeaderFieldsMiddlewareComponent.swift [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import Foundation
+
+final class HeaderFieldsMiddlewareComponent: MiddlewareComponent {
```
[the remark this answers is not in the document]
Comment (SimonNumberTwo) [reply Sources/Jetworking/Middleware/Defaults/HeaderFieldsMiddlewareComponent.swift:3]: See comment above. Accessibility will be adjusted.

File: Sources/Jetworking/Middleware/Defaults/AuthenticationMiddlewareComponent.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (JensK611) [reply Sources/Jetworking/Middleware/Defaults/AuthenticationMiddlewareComponent.swift]: Ok, I think we should keep the method out of the component.

I understand your point with preventing overriding, but why should it be forbidden? For example if we dynamically add header fields, we could just subclass `HeaderFieldsMiddlewareComponent´ and adjust the `headerFields` property. Whats your suggestion to solve the dynamic header fields?

[the remark this answers is not in the document]
Comment (SimonNumberTwo) [reply Sources/Jetworking/Middleware/Defaults/AuthenticationMiddlewareComponent.swift]: The dynamic header fields problem is already solved by setting them when initialising the client configuration (see Testcases). 

In my opinion we should take care about licensing here as well as the warranty which comes with it. The idea behind these components is to have a default working set which can be exchanged by writing own ones. 

Keep in mind when possible problems occur and someone mentions the default component but has adjusted it in some way we are not familiar with and we are searching for an error which is not within this code base we might get stuck or use a lot of time to find out it is not due to our code. 

Summing up these thoughts I would keep it `final` and one then can still copy the contents of these default components and just add their customisations by simply conforming to the `RequestMiddlewareComponent` protocol.

[the remark this answers is not in the document]
Comment (SimonNumberTwo) [reply Sources/Jetworking/Middleware/Defaults/AuthenticationMiddlewareComponent.swift]: The `AuthenticationMethod` enum is now within the `AuthenticationRequestMiddlewareComponent` as it is nowhere else used.

File: Sources/Jetworking/Middleware/Defaults/AuthenticationMiddlewareComponent.swift [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            let authorizationValue = authorizationHeader.values.first
+        {
+            mutatedRequest.addValue(authorizationValue, forHTTPHeaderField: authorizationKey)
+        }
+        
+        return mutatedRequest
+    }
+    
+    private func getAuthorizationHeader() -> [String: String]? {
+        let authorizationHeaderKey: String = "Authorization"
```
[the remark this answers is not in the document]
Comment (SimonNumberTwo) [reply Sources/Jetworking/Middleware/Defaults/AuthenticationMiddlewareComponent.swift]: Added `Constants` within `AuthenticationRequestMiddlewareComponent` with all hard coded strings.

File: Sources/Jetworking/Middleware/Defaults/AuthenticationMiddlewareComponent.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (JensK611) [reply Sources/Jetworking/Middleware/Defaults/AuthenticationMiddlewareComponent.swift]: Hm ok. But it is not such dynamic as I mean. Sometimes it is necessary to change header while using the same configuration. For example after a login. What do you think about passing a closure which is returning a dictionary of headers? This could be `@autoclosure` to achieve the functionality of your example?

Some thing like in this way:

```swift
import Foundation

public final class HeaderFieldsRequestMiddlewareComponent: RequestMiddlewareComponent {
    private var headerFields: () -> [String: String]

    init(headerFields: @escaping @autoclosure (() -> [String: String])) {
        self.headerFields = headerFields
    }

    public func process(request: URLRequest) -> URLRequest {
        var mutatedRequest: URLRequest = request
        headerFields().forEach { key, value in
            mutatedRequest.addValue(value, forHTTPHeaderField: key)
        }

        return mutatedRequest
    }
}
```

The advantage would be that you can pass a static dictionary or a closure which is executed for each request.

[the remark this answers is not in the document]
Comment (SimonNumberTwo) [reply Sources/Jetworking/Middleware/Defaults/AuthenticationMiddlewareComponent.swift]: Added autoclosure functionality to be highly flexible.

File: Sources/Jetworking/Middleware/Defaults/HeaderFieldsMiddlewareComponent.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (SimonNumberTwo) [reply Sources/Jetworking/Middleware/Defaults/HeaderFieldsMiddlewareComponent.swift]: See discussion above.
